OpenRA.Mods.Common/Traits/Mobile.cs

```diff
@@ -186,36 +194,48 @@ public bool CanEnterCell(World world, Actor self, CPos cell, Actor ignoreActor =
 		// Determines whether the actor is blocked by other Actors
 		public bool CanMoveFreelyInto(World world, Actor self, CPos cell, Actor ignoreActor, CellConditions check)
 		{
+			if (!check.HasCellCondition(CellConditions.TransientActors))
+				return true;
+
 			if (SharesCell && world.ActorMap.HasFreeSubCell(cell))
 				return true;
 
-			if (check.HasFlag(CellConditions.TransientActors))
-			{
-				var canIgnoreMovingAllies = self != null && !check.HasFlag(CellConditions.BlockedByMovers);
-				var needsCellExclusively = self == null || Crushes == null || !Crushes.Any();
-				foreach (var a in world.ActorMap.GetUnitsAt(cell))
-				{
-					if (a == ignoreActor)
-						continue;
-
-					// Neutral/enemy units are blockers. Allied units that are moving are not blockers.
-					if (canIgnoreMovingAllies && self.Owner.Stances[a.Owner] == Stance.Ally && IsMovingInMyDirection(self, a)) continue;
-
-					// Non-sharable unit can enter a cell with shareable units only if it can crush all of them.
-					if (needsCellExclusively)
-						return false;
-					var crushables = a.TraitsImplementing<ICrushable>();
-					if (!crushables.Any())
-						return false;
-					foreach (var crushable in crushables)
-						if (!crushable.CrushableBy(Crushes, self.Owner))
-							return false;
-				}
-			}
+			foreach (var otherActor in world.ActorMap.GetUnitsAt(cell))
+				if (IsBlockedBy(self, otherActor, ignoreActor, check))
+					return false;
 
 			return true;
 		}
 
+		bool IsBlockedBy(Actor self, Actor otherActor, Actor ignoreActor, CellConditions check)
+		{
+			// We are not blocked by the actor we are ignoring.
+			if (otherActor == ignoreActor)
+				return false;
+
+			// If the check allows: we are not blocked by allied units moving in our direction.
+			if (!check.HasCellCondition(CellConditions.BlockedByMovers) &&
+				self != null &&
+				self.Owner.Stances[otherActor.Owner] == Stance.Ally &&
+				IsMovingInMyDirection(self, otherActor))
+				return false;
+
+			// If we cannot crush the other actor in our way, we are blocked.
+			if (self == null || Crushes == null || Crushes.Length == 0)
+				return true;
+
+			// If the other actor in our way cannot be crushed, we are blocked.
+			var crushables = otherActor.TraitsImplementing<ICrushable>();
+			if (!crushables.Any())
+				return true;
```

Comment:
I prefer to group it since it's all related to crushability and the comment applies to the whole chunk of code, but I'll add a newline if that's the preference (we should try and find a StyleCop rule that does that, if it is indeed supposed to be the convention).